### ISSUE TITLE
Row names

### ISSUE THREAD

**Issue by LJWatson** [w3c/csvw#549]
Tabular Data Model.

In 4.3 Columns, a name is described as one of the core annotations. In 4.4 Rows, there is no such core annotation.

It's (quite) possible I don't understand the purpose of the name annotation, but if it is what would be converted into the column header (like the th in an HTML table for example), it is equally important that something similar exists for each row.

Screen reader users cannot glance at the row/column headers in a table, and so need some way of being told which row/column the cell they're currently focused on relates to. This is done by announcing the column header as focus moves left/right through a row, or the row header as focus moves up/down through a column, followed by the cell content.

If the name annotation is intended to be used in this way, the row name may be as important in accessibility terms as the column name. If I have misunderstood the purpose of the name annotation, feel free to close this issue with my apologies.

**Comment by chaals** [comment]
I noted essentially the same issue in my (omnibus, sorry) comments.

Having an annotation property for a row, which describes how to generate an "accessible name" - the default might be the row number, but it is equally likely that there is a column or a combination of cells from more than one column whose content is a human-readable label relevant to the row in question...

**Comment by danbri** [comment]
I may be missing something here, my guess is that this apparent asymmetry comes from our focus on tabular data in standalone CSV files (or Tab-separated). Although we've tried to leave the door open for bindings to a wider range of formats (including HTML tables, spreadsheet representations etc.), the central stereotypical use case have been the "blahblah.csv" file which most commonly have headers at the top, and very rarely per-row.

**Comment by chaals** [comment]
Yes, it is pretty clear that often the "name" we're looking for will be nothing more than a row number.

But often there will be something better. In your examples of trees for pruning, the address of the tree would be a more useful human-readable identifier. For a set of class records, a student's name. etc etc.

The asymmetry exists in the sense that most data won't have defined something useful. But it would be helpful to allow for the cases where there is useful row data and people bother to provide it - especially as a user-defined option to a processor:

parse --rownames "%3, %2" students.csv

or something.

Dealing with row numbers is a pretty tough cognitive load, labelling things is really helpful. Hence the common feature of labelling at least columns…

**Comment by danbri** [comment]
Perhaps the csv2rdf aspect could help, as it lets you say "this column maps into dc:title" etc., i.e. for each row the cell values can be associated with better known properties.  Rather than giving a fixed name for each row, you get hints that can help you understand how to compose a short (or long) summary of the item described by that row. An advantage of this approach is that it works out-of-band with existing (CSV etc) tables that don't offer explicit per-row names.

**Comment by chaals** [comment]
"Rather than giving a fixed name for each row, you get hints that can help you understand how to compose a short (or long) summary of the item described by that row"

Yes, that's about what I was thinking the common pattern would be in practice.

But rather than tie it to one of the many possible RDF naming properties, it seems reasonable to define it as a base annotation property. it isn't like they have to have a value…

**Comment by iherman** [comment]
Discussed at telco 2015.05.13, see  http://www.w3.org/2015/05/13-csvw-irc#T14-37-06

**Comment by gkellogg** [comment]
RESOLVED: create a row name annotation, which is not set through metadata. If not set, it’s value is “_row.[N]” in standard mode output, row name is included for each row

**Comment by gkellogg** [comment]
The metadata document doesn't currently have a way to set the row name annotation, but the door is open for other mechanisms, or future versions of the metadata document, to create such annotations.

**Comment by 6a6d74** [comment]
@gkellogg - when you have updated the model document (and namespace doc) let me know and I will update the CSV2\* conversion docs.

**Comment by LJWatson** [comment]
@iherman @gkellogg 

The proposal looks good. Is there no way to introduce the ability to set a row name annotation in the metadata document, without waiting for a future revision though?

**Comment by gkellogg** [comment]
@LjWatson there is not currently a Row class to which we might attach a name property. Note that for column names, we can infer that from a column title or column number. Introducing a row name would best come from a Row title, for which we had a concept (headerColumns) which was removed. Adding this now would be a pretty big change. Note, however, that in serializations, you can use cell values when constructing URLs, which serves much the same purpose.

**Comment by JeniT** [comment]
I believe that what @LjWatson and @chaals are asking for are more akin to row _titles_ than row _names_ (at least if we're keeping the terminology consistent across columns and rows). A column _name_ is a _machine-readable_ identifier for the column, used when referring to the column in eg URL templates or foreign keys. A column _title_ is a human-readable label for the column, which can be provided in multiple languages.

If that's correct, the row _primary key_ is the nearest equivalent to a name for a row. If defined, it indicates the column values that, when combined, provide an unique identifier for that row (eg you might have a primary key that used the first name & last name columns in the table to uniquely identify the row). It seems like this could be a good thing for a screen reader to use to identify the row, and we could add some text recommending that they use the primary key rather than the row number when a primary key is available. Would that be helpful?

We could also introduce a similar property to `primaryKey` to identify the columns that provide human-readable titles for the row (which might be available in different languages in some circumstances), eg `rowTitles` as a column reference property (ie referencing one or more columns). This could populate a new _titles_ annotation on rows in the model. We could add text recommending using those titles rather than the primary key or row number, when available.

We can also add similar text around the use of _titles_, _name_ and _number_ for identifying columns.

Does that sound like a good approach?

**Comment by gkellogg** [comment]
I'll hold off editor action until this is clarified; @JeniT, if you don't mind, I'll give you the action.

**Comment by iherman** [comment]
> On 14 May 2015, at 21:26 , Jeni Tennison notifications@github.com wrote:
> 
> I believe that what @LjWatson and @chaals are asking for are more akin to row titles than row names (at least if we're keeping the terminology consistent across columns and rows). A column name is a machine-readable identifier for the column, used when referring to the column in eg URL templates or foreign keys. A column title is a human-readable label for the column, which can be provided in multiple languages.

I think you are right.

> If that's correct, the row primary key is the nearest equivalent to a name for a row. If defined, it indicates the column values that, when combined, provide an unique identifier for that row (eg you might have a primary key that used the first name & last name columns in the table to uniquely identify the row). It seems like this could be a good thing for a screen reader to use to identify the row, and we could add some text recommending that they use the primary key rather than the row number when a primary key is available. Would that be helpful?
> 
> We could also introduce a similar property to primaryKey to identify the columns that provide human-readable titles for the row (which might be available in different languages in some circumstances), eg rowTitles as a column reference property (ie referencing one or more columns). This could populate a new titles annotation on rows in the model. We could add text recommending using those titles rather than the primary key or row number, when available.

I am actually more in favour of using rowTitles explicitly, rather than relying on primary keys only. Conceptually, primary keys have a more technical nature and goal, it may not be a good idea to mix those. Of course, visualizers/screen readers may use primary keys (or simply row numbers) but I believe our recommendation ('recommendation' not 'Recommendation') should be for authors to use a separate rowTitles for larger, more complex tables to make things explicit.

The discussion on the call went one step further, bringing in the RDF mapping. What we thought of doing is that, if there is a name annotation on the row then, in standard mode, that annotation also generates an equivalent triple in the RDF output for that specific row. The question is whether we should do that for rowTitles? In the absence of rowTitles, should we do it for primary Keys? I think whatever we do for RDF should be consistent with the way we formulate the general recommendation.

Ivan

> We can also add similar text around the use of titles, name and number for identifying columns.
> 
> Does that sound like a good approach?
> 
> —
> Reply to this email directly or view it on GitHub.

---

Ivan Herman, W3C
Digital Publishing Activity Lead
Home: http://www.w3.org/People/Ivan/
mobile: +31-641044153
ORCID ID: http://orcid.org/0000-0003-0782-2704

**Comment by LJWatson** [comment]
@iherman

A human readable title is the ideal outcome. When looking at tabular data with a screen reader, it is possible to move up/down through columns, or left/right through rows.

When focus is somewhere in the middle of a table, a sighted person will glance at the row/column header to orientate themselves and remind themselves which row/column they’re in. This outcome is handled in screen readers by announcing the column header as you move focus into a cell from left/right, or the row header as you move focus into a cell from up/down.

Where the spec has an annotation that specifically denotes the column name/title/header, we’re after the same thing for a row header. A key would do, but something human readable is always going to be more helpful.

Thanks for the time and consideration you’re giving this.

**Comment by chaals** [comment]
Yes, I think Jeni's suggestion of rowTitle is about what we are after - human-readable titles may not be the same as a single cell's content. E.g. A table including cells for "first name" "other names" "family name" … might use a combination of first name and family name as a useful row title, or a table of timezone data that has "continent", "city", "base offset", "summertime starting formula", "summetime ending formula", ... might use continent+"/"+city as a title…

In my "dream up a solution in a minute" approach, I'd have a substitutable syntax for it, a bit like printf() where you could use the contents of cells, as well as straight text to combine them.

**Comment by danbri** [comment]
@chaals your 'printf' is close to a design area we looked into the csv2rdf and csv2json pieces, use of templates e.g. in the Mustache style. We chose not to go there for 1.0 as we didn't feel we were in a position to adapt/spec/design that work well enough with the time and team available. However the specs are designed so that this kind of templating (alongside the likes of R2RML for declarative mappings into RDF) can be layered on later. If and when someone actually does this, there could be a good case made for standardization; but for now there's little more than a few rough prototypes around.
